### ISSUE TITLE
Evolving Perspectives in Image Style Transfer: Exploring Model Capabilities and Commercial Applications

### ISSUE THREAD

**Issue by YueTang25** [YueTang25/deep-learning.github.io#4]
Authors: Yue Tang, Tianfang Fang

## Background
Image Style Transfer is a significant research direction in the field of computer vision, which aims to apply the artistic style of one image to another content image. With the advancement of deep learning technologies, numerous models have been developed for efficient and high-quality style transfer. The rise of multi-modal models and large language models has greatly enhanced models' understanding of content, leading to considerable changes in the performance of image style transfer tasks. In this blog, we will begin with the VGG model and conclude with the multimodal model BLIP, illustrating the advancements in deep learning models' content understanding through the task of image style transfer. We will also discuss the varying performances of different models.

## Neural Style Transfer Based on Convolutional Neural Networks: The Case of the VGG Model
In the early stages of image style transfer, researchers used CNN-based models to perform the task. One of the pioneering ideas was proposed by Leon A. Gatys in 2015, who demonstrated the possibility of using the VGG model for style transfer tasks. Here, we introduce the VGG model and its application in the task of image style transfer.

For further details on Gatys' contribution, please refer to his seminal paper: Gatys, L. A., Ecker, A. S., & Bethge, M. (2015). A Neural Algorithm of Artistic Style. arXiv preprint arXiv:1508.06576.

### VGG Model Structure
The VGG model was previously used for classic computer vision tasks such as training image classifiers. Its distinctive feature is the use of small convolutional kernels and a deeper network structure. A notable improvement of VGG16 over AlexNet is the use of multiple 3x3 convolutional kernels in succession instead of the larger kernels used in AlexNet (11x11, 7x7, 5x5). For a given receptive field (the local size of the input image related to the output), using stacked small kernels is superior to using larger ones. This is because multiple non-linear layers can increase the depth of the network, enabling it to learn more complex patterns while being more cost-effective (using fewer parameters). In simpler terms, in VGG, three 3x3 convolutional kernels are used to replace a single 7x7 kernel, and two 3x3 kernels are used in place of a 5x5 kernel. The main purpose of this design is to increase the network's depth while maintaining the same receptive field, thereby enhancing the effectiveness of the neural network to some extent.

For further details on the VGG model, please refer to the original paper: Simonyan, K., & Zisserman, A. (2014). Very Deep Convolutional Networks for Large-Scale Image Recognition. arXiv preprint arXiv:1409.1556.

### Applying the VGG Model to Style Transfer
Leon A. Gatys introduced the idea of using the VGG model for style transfer in 2015 through two key studies: "Texture Synthesis Using Convolutional Neural Networks" and "Image Style Transfer Using Convolutional Neural Networks." In his research, Gatys utilized the VGG model to analyze both the style image and the content image separately, and then performed style transfer on the content image. The fundamental concept involves extracting "texture-like" elements from the style image and superimposing them onto the content image. The process is illustrated in the following figure.
<div align="center">
<img src="https://github.com/user-attachments/assets/883a38dc-19e4-40c9-8012-06c4368f6d3d" width="550" height="300">
<img src="https://github.com/user-attachments/assets/fcbdf271-9862-4160-ab7e-42056459ce23" width="500" height="500">
</div>

This style transfer technique has a clear drawback: it is difficult for the model to discern which elements of a style image pertain to content and which to style. Therefore, even in the author's examples, where the chosen style images are characterized by strong styles and relatively simple content (such as the works of Van Gogh and other abstract artists), we can still vaguely see elements like railings being preserved in the resulting images.

Moreover, this method aims to maintain content stability through invariant values in the higher layers of the image, which does not truly comprehend the distinction between style and content. Consequently, the results can be subpar on some images. For instance, when using a photograph of Audrey Hepburn and a starry night sky for training, the expected painterly style photo was not achieved; instead, the result was somewhat unusual:
<div align="center">
<img src="https://github.com/user-attachments/assets/9258d8cf-6401-4f90-886b-ab77030a9e99" width="800" height="150">
</div>

On one hand, this approach offers a promising direction for style transfer tasks. On the other hand, the shortcomings in the model's performance further illustrate that there is a need to enhance its ability to understand content. This dual perspective not only highlights current achievements but also underscores the ongoing challenges in advancing the capabilities of neural networks in artistic applications.

## Transformer-Based Style Transfer: The Case of StyTR-2
In the realm of NLP, the advent of attention mechanisms and transformers represents a significant leap in content understanding capabilities. Transformer-based models, including BERT and LLAmA, have demonstrated improved performance in understanding content. Researchers have also found that transformers perform exceptionally well in the field of computer vision (CV). Traditional CNN methods, due to their locality and spatial invariance, struggle to extract and maintain the global information of input images, which can lead to suboptimal results in artistic rendering. To address these challenges, the StyTR-2 method was proposed, employing transformers for unbiased image style transfer.

The StyTR-2 method incorporates two different transformer encoders designed specifically for generating domain-specific sequences for content and style, respectively. Following the encoders, a multi-layer transformer decoder styles the content sequence according to the style sequence. Additionally, the authors have critiqued existing position encoding methods, proposing a scale-invariant and content-aware position encoding (CAPE) that is more suited to image style transfer tasks. Figure 1 illustrates the structure of StyTR-2.
![image4](https://github.com/user-attachments/assets/e864a2b8-2b9f-43da-b684-1d2e121b12c3)
![image5](https://github.com/user-attachments/assets/b751e22e-66d9-40b7-8584-3aac12fc3c61)

Results Using the Same Content and Style Images:
<div align="center">
<img src="https://github.com/user-attachments/assets/b7d5c915-d395-4a0d-8cf0-3e9bb6e456d5" width="600">
<img src="https://github.com/user-attachments/assets/280a5fa0-2f68-4394-8808-a9699e8cd272" width="600">
</div>
As demonstrated by the results starting with StyTR-2, the model can now completely separate the content of the style image. From our results, it is evident that the newly generated image retains all the content elements, such as Audrey Hepburn's earrings and the details of the house. In terms of the style image, it preserves none of the content but maintains aspects such as color, texture, and brushstrokes without retaining any shapes.
In comparison, StyTR-2 acts more like applying an advanced "filter" to the image, making it easier to extract and transfer a style based more on color and texture. This approach also avoids the uncontrollable outcomes associated with VGG model style transfers, which can result in very odd-looking photos.

## Style Transfer Based on Generative Adversarial Networks: The Case of CycleGAN
As evident from the discussion above, the models described so far involve style transfer based on paired images—providing a content image and a style image for training. This approach has a clear drawback: a single style image may not adequately represent a style. For instance, if we wanted to describe Van Gogh's style, what would that include? High saturation of colors? Harmonious color schemes? Natural brush strokes? All of these elements are correct, but 'Starry Night' does not represent 'Wheatfield with Crows'. If we could incorporate all his paintings into the training set and extract the common elements among these works, we would undoubtedly gain a better understanding of his style. CycleGAN offers a solution to this challenge.

Published in the ICCV 2017 paper "Unpaired Image-to-Image Translation using Cycle-Consistent Adversarial Networks," CycleGAN has been utilized for image style transfer tasks. Unlike previous GANs that generated images in a single direction, CycleGAN breaks through the Pix2Pix limitation of requiring one-to-one correspondence between dataset images by using a bidirectional cyclical generation structure, hence the name CycleGAN.

CycleGAN operates as a GAN model, engaging in adversarial training between discriminators and generators to learn the pixel probability distribution of dataset images for generating new images. To achieve style transfer from domain X to domain Y, the GAN network must fit the pixel probability distribution of domain Y while preserving the characteristics of domain X. The sets of images from domains X and Y inputted into CycleGAN do not have a one-to-one correspondence. Even if we input domain X images as constraints into a CGAN, it would not effectively limit the model to retain features from domain X because the paired sets of images are randomly matched, and CGAN cannot learn any connection between them. Therefore, CycleGAN employs a brilliant design: it adds "cyclical generation" and optimizes a consistency loss to replace the constraint conditions used in CGAN, allowing the generator to retain original domain features without the need for one-to-one correspondence between training set images.
<div align="center">
<img src="https://github.com/user-attachments/assets/633bb060-49e1-4ab4-815c-3c5d746b9bb6" width="600">
</div>

#### Effectiveness of Style Transfers Across Various Styles
According to the examples provided by the authors, the images demonstrate commendable style transfer results across various styles. The ability of CycleGAN to maintain the essential features of the domain while successfully adopting the stylistic elements of another has been effectively showcased. These examples illustrate not only the versatility of CycleGAN in handling different artistic styles, from classical to modern and abstract, but also its proficiency in capturing and translating the distinctive characteristics of each style without needing a direct one-to-one correspondence between the source and target images.
![image9](https://github.com/user-attachments/assets/a934f612-04dc-481c-864d-d33f6b6ab543)

Results Using the Same Content and Style Images (real image -> Monet style -> Vangogh style):
<div align="center">
<img src="https://github.com/user-attachments/assets/fa64af03-fffb-4d6c-9a3f-da4d82461f80" width="700">
<img src="https://github.com/user-attachments/assets/b270f207-befd-4e00-9704-6ab1debfb1fd" width="600">
</div>

Additionally, we conducted style swaps between the works of Van Gogh and Monet, with the results as follows (real image -> processed image):
<div align="center">
<img src="https://github.com/user-attachments/assets/dda05398-bf36-4ca2-84f8-601b23514286" width="400">
</div>
From the results observed, it is evident that CycleGAN has improved its understanding of styles and is no longer influenced by the extreme colors of a single artwork. By moving away from paired training, the model has developed a deeper understanding of styles.

## Multimodal Style Transfer Model: The Case of StyleCLIP
With the introduction of the CLIP model, multimodal models have become increasingly popular, opening up new possibilities for style transfer tasks. In 2021, StyleCLIP was introduced as a groundbreaking style transfer model. StyleCLIP allows users to make more specific and varied changes to images through textual instructions.

For example, as demonstrated in the paper, StyleCLIP can perform complex adjustments to images effectively. These adjustments include removing a beard, changing hairstyles, and other significant modifications that go beyond traditional style transfer capabilities, leveraging the intersection of text and image modalities to guide the transformation process.
![image10](https://github.com/user-attachments/assets/3b90acc9-1677-4dde-8fee-ecdba983e844)

This undoubtedly offers greater freedom and more possibilities in style transfer. Below is an example where we used StyleCLIP to change Audrey Hepburn's hair color:
<div align="center">
<img src="https://github.com/user-attachments/assets/7ee22893-b79d-43c4-af7e-b43551acd06a" width="600">
</div>

## Innovative Practices in the Industry
After discussing the performance of various models in image style transfer tasks, we can now turn our attention to more industrial and practical questions: Can these models, directly or through retraining, become a new feature on social media platforms? While it's possible, the results might not be as good as expected.

For real users, being fun is a crucial metric for evaluating apps, which includes making changes that are interesting and ensuring that the results do not degrade the original photos. Whether there is significant content loss might not be of much concern to actual users.

This year's breakout photo app, which can transform photos into claymation-style, makes images appear as though they're from a stop-motion animation set, which is quite fascinating. If we examine several photos, we'll notice that this style transformation generally does not preserve facial features of the subjects; it usually just selects facial features from a limited set of patterns that are somewhat similar. However, this does not detract from its success as a feature in a photo app.
<div align="center">
<img src="https://github.com/user-attachments/assets/46b82158-c4c5-4163-a8ef-f5fabca1e3fd" width="550" height="300">
<img src="https://github.com/user-attachments/assets/7754ff04-add1-4621-afab-a39f5d2cdb4c" width="350" ><img src="https://github.com/user-attachments/assets/97454601-a399-4647-82d6-4a6fab359a41" width="350">
</div>

In my experiments, I also attempted to transform my Ragdoll cat into the style of a gold foil painting. From the transformation results, sometimes my Ragdoll cat would appear more like a Siamese cat, but the generated photos were always of cute kitties.
![image14](https://github.com/user-attachments/assets/e357a0ea-bd07-4572-9b91-60d77f43ff18)
![image15](https://github.com/user-attachments/assets/735dad7b-3753-437e-becc-c55a103b7006)

From what can be observed, these transformations likely involve first extracting the content embeddings of the photographs and then blurring or abstracting these embeddings. The style image is then generated based on these processed embeddings. Perhaps it is the discarding of certain content details that allows the generated images more flexibility to better align with the target style. This trade-off makes the photos interesting and engaging. Such tasks, while perhaps closer to a special effect than traditional style transfer, undoubtedly represent an attractive application within the field of computer vision.

Reference Code:
1. https://github.com/diyiiyiii/StyTR-2
2. https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix
3. https://github.com/orpatashnik/StyleCLIP
